src/common/libjob/spec.c

```diff
@@ -0,0 +1,337 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <errno.h>
+#include <jansson.h>
+#include <string.h>
+
+#include "specutil.h"
+#include "src/common/libutil/errno_safe.h"
+
+#include "spec.h"
+
+typedef struct flux_jobspec1 {
+    json_t *jobspec_json;
+} flux_jobspec1_t;
+
+/* Set the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * 'json_str' should be a json-encoded string.
+ * Return 0 on success, -1 otherwise with errno set.
+ */
+int flux_jobspec1_attr_set (flux_jobspec1_t *jobspec,
+                            const char *path,
+                            const char *json_str)
+{
+    json_error_t error;
+    json_t *value;
+    int rc;
+
+    if (!(value = json_loads (json_str, JSON_DECODE_ANY, &error))) {
+        errno = EINVAL;
+        return -1;
+    }
+    rc = specutil_attr_set (json_object_get (jobspec->jobspec_json, "attributes"),
+                            path,
+                            value);
+    ERRNO_SAFE_WRAP (json_decref, value);
+    return rc;
+}
+
+/* Get the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * Store the result as a json-encoded string in 'json_str'. The string must be
+ * released with 'free()'. Note that the string may not be a valid JSON document,
+ * since it may (for instance) define a number or string rather than an array or object.
+ * Return 0 on success, -1 otherwise.
+ */
+int flux_jobspec1_attr_get (flux_jobspec1_t *jobspec, const char *path, char **json_str)
```

Comment:
I don't have any well-defined use cases other than the "stop-tasks-in-exec" I mentioned above, so I think replacing it is a good idea (ditto for your comment about ``attr_set``). For the average user it's probably easier to use ``attr_unpack`` rather than to build JSON strings.